[vLLM] stablehlo.gather lowered to ttnn.embedding (not ttnn.gather), corrupting integer vocab IDs in non-greedy sampling

**Status update:** The topk index bug described below was a test methodology error (topk ordering is non-deterministic — see latest comment). The confirmed root cause is the `stablehlo.gather → ttnn.embedding` lowering. See the [latest comment](https://github.com/tenstorrent/tt-xla/issues/4329#issuecomment-4285626447) for full findings, tests, and fix status.

---

(Under debug, not proven yet)

--

Two TT device bugs were found while debugging non-greedy vLLM sampling garbage output. Not yet fully proven (opening to track, not lose the repro). Both block correct non-greedy device sampling when using the chunked-topk approach from `a014385f0`.

**Reproducer:** `tests/torch/ops/test_tt_sampling_bugs.py` on branch `kmabee/vllm_perf_apr16_roll_back` (all 6 assertions fail). Cherry pick 7b1226ee4 if you want to avoid switching to my branch - Full analysis in `vllm_sampling_tt_bugs.md` on the same branch/commit.

cmd:

```
pytest -svv tests/torch/ops/test_tt_sampling_bugs.py |& tee test_tt_sampling_bugs.log

============================================================================================================================================================================== short test summary info ==============================================================================================================================================================================
FAILED tests/torch/ops/test_tt_sampling_bugs.py::test_topk_index_correctness[32768-k32-vllm-chunk] - AssertionError: topk index mismatch: 0.4688 (15/32 correct) — TT runtime bug
assert 0.46875 > 0.99
FAILED tests/torch/ops/test_tt_sampling_bugs.py::test_topk_index_correctness[32768-k64] - AssertionError: topk index mismatch: 0.3281 (21/64 correct) — TT runtime bug
assert 0.328125 > 0.99
FAILED tests/torch/ops/test_tt_sampling_bugs.py::test_topk_index_correctness[16384-k32] - AssertionError: topk index mismatch: 0.6250 (20/32 correct) — TT runtime bug
assert 0.625 > 0.99
FAILED tests/torch/ops/test_tt_sampling_bugs.py::test_topk_index_correctness[65536-k32] - AssertionError: topk index mismatch: 0.5312 (17/32 correct) — TT runtime bug
assert 0.53125 > 0.99
FAILED tests/torch/ops/test_tt_sampling_bugs.py::test_gather_int64_correctness[opt125m] - AssertionError: gather int64 mismatch: cpu=2814 dev=2816 (local_idx=58) — TT runtime bug
assert 2814 == 2816
FAILED tests/torch/ops/test_tt_sampling_bugs.py::test_gather_int64_correctness[llama] - AssertionError: gather int64 mismatch: cpu=91715 dev=91648 (local_idx=68) — TT runtime bug
assert 91715 == 91648
========================================================================================================================================================================== 6 failed, 18 warnings in 5.09s ===========================================================================================================================================================================
```

## Bug 1: `torch.topk` returns correct values but wrong indices

`torch.topk(x, k=32, dim=-1)` on a `[1, N]` float32 tensor where N is a power of 2 (tested: 16384, 32768, 65536) returns correct VALUES (cosine sim = 1.000 vs CPU) but only 30–65% correct INDICES. This is the signature of a bitonic sort that sorts values correctly but has broken index tracking — values and indices are swapped together so values land in correct ranked positions while indices go to wrong positions. Related to known issue #1797 (https://github.com/tenstorrent/tt-metal/issues/37571)

Observed index match rates:
- `[1, 32768], k=32`: 47% (15/32 correct)
- `[1, 32768], k=64`: 33% (21/64 correct)
- `[1, 16384], k=32`: 63% (20/32 correct)
- `[1, 65536], k=32`: 53% (17/32 correct)

## Bug 2: `torch.gather` on int64 tensors returns wrong values

`gather(dim=1)` on an int64 tensor returns values with a small but non-deterministic offset vs CPU (observed: off by 2 and off by 67 in two test cases). Could be a gather kernel bug or a CPU→device transfer corruption for int64 — not yet determined which.

## Impact

Both bugs cause non-greedy vLLM device sampling to produce garbage token output. Greedy sampling is unaffected. Workaround: `cpu_sampling=True`.

## Uncertainty

Bug 1 may be in the `ttnn.topk` kernel or in the XLA lowering for these specific shapes. Bug 2 may be in the gather kernel or in how int64 tensors are transferred to TT device. Opening this issue to preserve the repro and evidence while investigation continues.

--- a/python_package/tt_torch/composite_ops.py
+++ b/python_package/tt_torch/composite_ops.py
@@ -270,6 +270,33 @@ def composite_scaled_dot_product_attention(
     return output
 
 
+def composite_gather(
+    input: Tensor,
+    dim: int,
+    index: Tensor,
+    *,
+    sparse_grad: bool = False,
+    out: Tensor | None = None,
+) -> Tensor:
+    """
+    Creates a composite gather operation for torch-xla using StableHLOCompositeBuilder.
+    Name of the composite op will be "tenstorrent.gather"
+
+    Args:
+        input: tensor to gather from
+        dim: axis along which to index
+        index: tensor with integer dtype which contains the indices to gather
+        sparse_grad: optional bool, if true the gradient computation will be sparse
+        out: preallocated output tensor, will be ignored when generating composite op
+    """
+    attrs = {"dim": dim, "sparse_grad": sparse_grad}
+    builder = StableHLOCompositeBuilder(name="tenstorrent.gather", attr=attrs)
+    input, index = builder.mark_inputs(input, index)
+    output = torch.gather(input, dim, index, sparse_grad=sparse_grad)
+    output = builder.mark_outputs(output)
+    return output
+
+
 ################# module replacements #################
 
 
@@ -459,6 +486,7 @@ replacements = {
         frozenset({0}): composite_topk_values,
         frozenset({1}): composite_topk_indices,
     },
+    torch.gather: composite_gather,
     # module replacements
     torch.nn.LayerNorm: replace_layer_norm_module,
     # TODO: uncomment once https://github.com/tenstorrent/tt-metal/issues/40916 is fixed
@@ -473,4 +501,5 @@ The mapped function must be a key in `replacements` for the rewrite to apply.
 """
 method_name_to_function = {
     "topk": torch.topk,
+    "gather": torch.gather,
 }

--- a/tests/torch/ops/test_composite_ops.py
+++ b/tests/torch/ops/test_composite_ops.py
@@ -15,6 +15,7 @@ from infra.evaluators import TorchComparisonEvaluator
 from infra.utilities.types import Framework
 from torch.nn import functional as F
 from tt_torch.composite_ops import (
+    composite_gather,
     composite_gelu,
     composite_group_norm,
     composite_layer_norm,
@@ -454,6 +455,72 @@ def test_patched_topk_both(input_shape, k):
     run_graph_test(
         TopKBoth(k),
         [input],
+        comparison_config=ComparisonConfig(),
+        framework=Framework.TORCH,
+        torch_options=options,
+    )
+
+
+@pytest.mark.nightly
+@pytest.mark.single_device
+@pytest.mark.parametrize(
+    "input_shape, dim",
+    [((4, 10), 0), ((4, 10), 1), ((2, 3, 8), 2)],
+)
+def test_composite_gather(input_shape, dim):
+    class GatherModel(torch.nn.Module):
+        def __init__(self, dim):
+            super().__init__()
+            self.dim = dim
+
+        def forward(self, x, index):
+            return composite_gather(x, self.dim, index)
+
+    options = {"tt_enable_composite_ops": False}
+
+    input_tensor = torch.randn(*input_shape)
+    # Build an index tensor with the same number of dims, gathering 2 elements along dim
+    index_shape = list(input_shape)
+    index_shape[dim] = 2
+    index = torch.randint(0, input_shape[dim], index_shape)
+
+    with torch._inductor.config.patch({"inplace_buffers": False}):
+        run_graph_test(
+            GatherModel(dim),
+            [input_tensor, index],
+            comparison_config=ComparisonConfig(),
+            framework=Framework.TORCH,
+            torch_options=options,
+        )
+
+
+@pytest.mark.nightly
+@pytest.mark.single_device
+@pytest.mark.parametrize(
+    "input_shape, dim",
+    [((4, 10), 0), ((4, 10), 1), ((2, 3, 8), 2)],
+)
+def test_patched_gather(input_shape, dim):
+    """torch.gather patched via tt_enable_composite_ops → composite_gather selected."""
+
+    class GatherModel(torch.nn.Module):
+        def __init__(self, dim):
+            super().__init__()
+            self.dim = dim
+
+        def forward(self, x, index):
+            return torch.gather(x, self.dim, index)
+
+    options = {"tt_enable_composite_ops": True}
+
+    input_tensor = torch.randn(*input_shape)
+    index_shape = list(input_shape)
+    index_shape[dim] = 2
+    index = torch.randint(0, input_shape[dim], index_shape)
+
+    run_graph_test(
+        GatherModel(dim),
+        [input_tensor, index],
         comparison_config=ComparisonConfig(),
         framework=Framework.TORCH,
         torch_options=options,

--- a/tests/torch/ops/test_gather.py
+++ b/tests/torch/ops/test_gather.py
@@ -1,0 +1,65 @@
+# SPDX-FileCopyrightText: (c) 2026 Tenstorrent AI ULC
+#
+# SPDX-License-Identifier: Apache-2.0
+import numpy as np
+import pytest
+import torch
+import torch_xla.runtime as xr
+from infra import Framework, run_graph_test
+from infra.testers.compiler_config import CompilerConfig
+from torch import nn
+from torch_xla.distributed.spmd import Mesh
+
+
+class SimpleGather(nn.Module):
+    def __init__(self, dim):
+        super().__init__()
+        self.dim = dim
+
+    def forward(self, input, index):
+        return torch.gather(input, self.dim, index)
+
+
+# Note, 1D and >4D tensors are not supported by ttnn.gather. A workaround will be added
+# to tt-mlir in the future to support them.
+@pytest.mark.parametrize(
+    "input_shape,index_shape,dim",
+    [
+        # ---- 2D: all dims (positive and negative), varying index shapes ----
+        pytest.param((32, 64), (16, 64), 0, id="2d_dim0"),
+        pytest.param((8, 32), (8, 16), 1, id="2d_dim1"),
+        pytest.param((16, 32), (16, 32), 1, id="2d_dim1_same_shape"),
+        pytest.param((16, 32), (16, 64), 1, id="2d_dim1_larger"),
+        pytest.param((16, 32), (8, 16), 1, id="2d_dim1_both_smaller"),
+        pytest.param((32, 64), (16, 32), 0, id="2d_dim0_both_smaller"),
+        pytest.param((32, 64), (16, 64), -2, id="2d_negdim0"),
+        pytest.param((8, 32), (8, 16), -1, id="2d_negdim1"),
+        # ---- 3D: all dims, varying index shapes ----
+        pytest.param((4, 32, 64), (2, 32, 64), 0, id="3d_dim0"),
+        pytest.param((4, 32, 64), (4, 8, 64), 1, id="3d_dim1"),
+        pytest.param((4, 16, 64), (4, 16, 32), 2, id="3d_dim2"),
+        pytest.param((4, 16, 64), (4, 16, 128), 2, id="3d_dim2_larger"),
+        pytest.param((4, 16, 64), (2, 8, 32), 2, id="3d_dim2_all_smaller"),
+        pytest.param((4, 32, 64), (4, 8, 64), -2, id="3d_negdim1"),
+        pytest.param((4, 16, 64), (4, 16, 32), -1, id="3d_negdim2"),
+        # ---- 4D: all dims ----
+        pytest.param((2, 4, 32, 64), (1, 4, 32, 64), 0, id="4d_dim0"),
+        pytest.param((2, 4, 32, 64), (2, 2, 32, 64), 1, id="4d_dim1"),
+        pytest.param((2, 4, 32, 64), (2, 4, 16, 64), 2, id="4d_dim2"),
+        pytest.param((2, 4, 32, 64), (2, 4, 32, 16), 3, id="4d_dim3"),
+        pytest.param((2, 4, 32, 64), (2, 4, 32, 16), -1, id="4d_negdim3"),
+        pytest.param((2, 4, 32, 64), (1, 2, 16, 32), 2, id="4d_dim2_all_smaller"),
+    ],
+)
+def test_gather_replicated(input_shape, index_shape, dim):
+    xr.set_device_type("TT")
+
+    input = torch.randn(*input_shape, dtype=torch.bfloat16)
+    index_max = input_shape[dim]
+    index = torch.randint(0, index_max, index_shape, dtype=torch.int64)
+
+    run_graph_test(
+        SimpleGather(dim),
+        [input, index],
+        framework=Framework.TORCH,
+    )

